cpp/src/io/utilities/type_inference.cuh

```diff
@@ -0,0 +1,303 @@
+    }
+    // A date field can have either one or two '-' or '\'; A legal combination will only have one
+    // of them To simplify the process of auto column detection, we are not covering all the
+    // date-time formation permutations
+    else if (((dash_count > 0 && dash_count <= 2 && slash_count == 0) ||
+              (dash_count == 0 && slash_count > 0 && slash_count <= 2)) &&
+             colon_count <= 2) {
+      atomicAdd(&column_info->datetime_count, 1);
+    }
+  }  // for
+}
+
+/**
+ * @brief Constructs column type histogram for a given column string input `data`.
+ *
+ * @tparam ColumnStringIter Iterator type whose `value_type` is convertible to
+ * `thrust::tuple<device_span, string_view>`
+ *
+ * @param options View of inference options
+ * @param data JSON string input
+ * @param column_strings_begin The begining of an offset-length tuple sequence
+ * @param size Size of the string input
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ * @return A histogram containing column-specific type counters
+ */
+template <typename ColumnStringIter>
+cudf::io::column_type_histogram detect_column_type(inference_options_view const& options,
+                                                   cudf::device_span<char const> data,
+                                                   ColumnStringIter column_strings_begin,
+                                                   std::size_t const size,
+                                                   rmm::cuda_stream_view stream)
+{
+  constexpr int block_size = 128;
+
+  auto const grid_size = (size + block_size - 1) / block_size;
+  auto d_column_info   = rmm::device_scalar<cudf::io::column_type_histogram>(stream);
+  CUDF_CUDA_TRY(cudaMemsetAsync(
+    d_column_info.data(), 0, sizeof(cudf::io::column_type_histogram), stream.value()));
+
+  detect_column_type_kernel<<<grid_size, block_size, 0, stream.value()>>>(
+    options, data, column_strings_begin, size, d_column_info.data());
+
+  return d_column_info.value(stream);
+}
+
+/**
+ * @brief Detects data type for a given JSON string input `data`.
+ *
+ * @throw cudf::logic_error if input size is 0
+ * @throw cudf::logic_error if data type detection failed
+ *
+ * @tparam ColumnStringIter Iterator type whose `value_type` is convertible to
+ * `thrust::tuple<device_span, string_view>`
+ *
+ * @param options View of inference options
+ * @param data JSON string input
+ * @param column_strings_begin The begining of an offset-length tuple sequence
+ * @param size Size of the string input
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ * @return The detected data type
+ */
+template <typename ColumnStringIter>
+cudf::data_type detect_data_type(inference_options_view const& options,
+                                 device_span<char const> data,
+                                 ColumnStringIter column_strings_begin,
+                                 std::size_t const size,
```

Comment:
```suggestion
                                 std::size_t const size,
                                 size_type omission_null_count
```

Reply:
Just to make sure, `size_type` is preferred over `uint32_t` or `std::size_t` right?

Reply:
https://github.com/rapidsai/cudf/blob/e099e01c9b6ab8a2db5d5ee446b8843ee6199acc/cpp/src/io/utilities/column_type_histogram.hpp#L27-L36

I do notice the column histogram is using `size_type` for counter types as well.

Reply:
No strong preference from my end. But, as you noted, I had seen that `size_type` would play nice with the arithmetic where `omission_null_count` would be involved, so I thought that'd make our lives easier 🙂

Reply:
Anything resembling a row count or limited by row count should be `size_type`, AFAIK.